### ISSUE TITLE
[clang-tidy] bugprone-exception-escape reported for structs with std::string

### ISSUE THREAD

**Issue by michaelvlach** [llvm/llvm-project#54668]
Starting with LLVM-14 the very useful `bugprone-exception-escape` check became borderline useless:
```
struct S
{
    std::string name;
};
```
will yield:
```
error: an exception may be thrown in function 'S' which should not throw exceptions [bugprone-exception-escape]
```
Needless to say it's almost every class/struct so having `NOLINT` on all of them is not an options. Defining all special members for structs like these is not reasonable either so one has to simply disable the check completely which is a shame.

And it is wrong anyway. In the above it CANNOT throw an exception at all. The move constructor of `std::string` is `noexcept` as is the destructor. Even if they did throw for whatever reason it would call `std::abort` so it is of no concern of `S` either way. The situation is slightly more complicated with the move assignment `operator=(&&)` but that is not what is being complained about here. The `function S` must be referring to the move constructor unless the warning message is wrong as well.

**Comment by LebedevRI** [comment]
That would be troubling indeed, but can you post the reproducer?
https://godbolt.org/z/eE9EzbEnG doesn't exhibit it.

**Comment by michaelvlach** [comment]
I was finally able to get a repro and nail it down somewhat. It happens only on Windows and not on Linux. The following is the minimum with which I was able to get it with MSVC 17.1.2 and LLVM 14 on Windows 10:
```
#include <string>

struct S
{
    std::string name;
};

class C
{
public:
    C() = default;
    C(const C &other) = delete;
    C(C &&other) noexcept
    {
        *this = std::move(other);
    }
    ~C() = default;

    auto operator=(const C&) -> C& = delete;
    auto operator=(C &&other) noexcept -> C&
    {
        if (this != &other)
        {
            this->s = std::move(other.s);
        }
        
        return *this;
    }

private:
    S s;
};
```
My observations:

1. It requires user-defined move constructor. With defaulted one it does not happen.
2. Marking the move constructor `noexcept` does not help.
3. The error message sometimes pinpoint the `operator=` but at various points I was getting only the original `an exception may be thrown in function 'S' which should not throw exceptions` with no further breakdown (having to guess it was indeed the move constructor.

Speculation:

- As it seems to be Windows specific it might be related to MSVC STL. I am using latest 17.1.2 (VS 2022). It might be that either the `string` constructors are not marked `noexcept` (although that does not seem to help anyway) or their implementation is indeed not `noexcept`. That being said even if this was true and MSVC STL was at fault the fact that it propagates from a system header through auto-generated struct constructor to the user code is still a valid issue imho.

Full clang-tidy log from it:
```
20269 warnings generated.
C:\dev\clang-tidy\source.cpp:4:8: error: an exception may be thrown in function 'S' which should not throw exceptions [bugprone-exception-escape,-warnings-as-errors]
struct S
       ^
C:\dev\clang-tidy\source.cpp:4:8: error: an exception may be thrown in function 'operator=' which should not throw exceptions [bugprone-exception-escape,-warnings-as-errors]
C:\dev\clang-tidy\source.cpp:14:5: error: an exception may be thrown in function 'C' which should not throw exceptions [bugprone-exception-escape,-warnings-as-errors]
    C(C &&other) noexcept
    ^
C:\dev\clang-tidy\source.cpp:21:10: error: an exception may be thrown in function 'operator=' which should not throw exceptions [bugprone-exception-escape,-warnings-as-errors]
    auto operator=(C &&other) noexcept -> C&
         ^
Suppressed 20265 warnings (20265 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
4 warnings treated as errors
```

**Comment by michaelvlach** [comment]
Looking at the implementation of STL in MSVC this is the offending constructor:

https://github.com/microsoft/STL/blob/main/stl/inc/xstring#L2470

```
    _CONSTEXPR20 basic_string() noexcept(is_nothrow_default_constructible_v<_Alty>)
        : _Mypair(_Zero_then_variadic_args_t{}) {
        _Mypair._Myval2._Alloc_proxy(_GET_PROXY_ALLOCATOR(_Alty, _Getal()));
        _Tidy_init();
    }
```

The `_Zero_then_variadic_args_t` is harmless (https://github.com/microsoft/STL/blob/main/stl/inc/xmemory#L1352):
```
struct _Zero_then_variadic_args_t {
    explicit _Zero_then_variadic_args_t() = default;
};
```
The `_Alloc_proxy` has two implementations. One can throw and is not declared `noexcept` but if the macro `_GET_PROXY_ALLOCATOR` returns the `FakeAllocator` (which seems to be the case for this particular call) it resolves to `noexcept` overload that does nothing.

The `_Tidy_init()` is declared `noexcept` (https://github.com/microsoft/STL/blob/main/stl/inc/xstring#L4594) but internally it can throw:
```
const pointer _New_ptr = _Al.allocate(_BUF_SIZE + 1); // throws
```
Even if the `_Alloc_proxy` was not resolved to the `noexcept` overload the constructor of `basic_string` is still marked `noexcept` and no exception can ever escape that context as it would simply call `std::abort` if any was thrown. Is this perhaps leaking diagnostics from system headers to the user code? I mean technically it is correct in that the allocation inside of `basic_string` can throw and it shouldn't but that is not something a user of STL can possibly fix or do anything about.

**Comment by LebedevRI** [comment]
This is starting to sound like #54956 that i reported later.

**Comment by michaelvlach** [comment]
I agree! The issue is that `clang-tidy` has to dig into the call tree in order to give us diagnostics, which is fine. What is not fine is leaking the diagnostics to the user of something that is declared `noexcept`. Therefore in your example I would expect:

1. `calee` getting flagged for calling throwing function and not guarding against `std::abort` with `try {} catch(...) {}`
2. `caller` to be fine because it calls only non/throwing stuff

What is happening with MSVC's `std::string` is that either `_Alloc_proxy` or `_Al.allocate` getting flagged (which is in fact correct) but then this diagnostics propagates all the way up to the end user ignoring a) several layers of `noexcept` and b) system header boundaries.

This particular case started to happen to me in LLVM14 but even before that I was getting annoyed by this:
```
int foo() noexcept { try {} catch (,,,) {} return 1; }
static const int i = foo();
```
getting flagged by `cert-err58-cpp` (statics being initialized with throwing expression). Not only the `foo` cannot throw as it is `noexcept` but it cannot call `abort` either due to the `try / catch`. I am fairly sure it has the same root cause as this and #54956 in that there is something throwing in the bottom and it gets incorrectly propagated all the way up. Lack of full trace to the offending source also does not help.

**Comment by IGR2014** [comment]
Same problems. MSVC STL from ver 17.3 preview 5 and ClangTidy 14.0.5

**Comment by ngg** [comment]
I've submitted https://reviews.llvm.org/D134588 as a candidate fix.

**Comment by iradization** [comment]
Hi, I'm running clang-tidy on macOS project and encountered some false-positives about the bugprone-exception-escape topic. how can I know if this fix already published on official build ? and if so, which build should I take ?  thanks !

**Comment by igagis** [comment]
I assume, since the issue is still open, the fix is not yet ready.

**Comment by PiotrZSL** [comment]
https://reviews.llvm.org/D145865
